typing/ctype.ml

```diff
@@ -2220,13 +2220,54 @@ let expand_head_opt env ty =
 let is_principal ty =
   not !Clflags.principal || get_level ty = generic_level
 
+module Bound_vars : sig
+  type t
+
+  val empty : t
+  val of_list : type_expr list -> t
+  val union : t -> t -> t
+  val mem : t -> type_expr -> bool
+  val is_empty : t -> bool
+end = struct
+  (* It is rare to care about which variables have been bound. This
+     implementation prioritizes the common case of caring only about the
+     presence of variables. *)
```

Comment:
I'm unconvinced that this is actually a perf improvement. In the common case, the sets are empty, so the set operations would be fast anyways and the laziness would just add overhead. I'm going to kick off some benchmarks to check.